app/bundles/EmailBundle/Controller/AjaxController.php

```diff
@@ -285,4 +286,56 @@ public function getEmailDeliveredCountAction(Request $request, CacheProvider $ca
             'delivered'   => $deliveredCount,
         ]);
     }
+
+    public function heatmapAction(Request $request): JsonResponse
+    {
+        $emailId = (int) InputHelper::clean($request->query->get('id'));
+
+        /** @var EmailModel $model */
+        $model = $this->getModel('email');
+
+        $email       = $model->getEntity($emailId);
```

Comment:
Using DI is a bit cleaner. Using the InputHelper seems unnecessary when it is retyped to int anyway.
```suggestion
    public function heatmapAction(Request $request, EmailModel $emailModel): JsonResponse
    {
        $emailId = (int) $request->query->get('id');
        $email   = $model->getEntity($emailId);
```